turtle.js

```diff
@@ -1,373 +1,604 @@
+const _radToDeg = rad => rad * 180 / Math.PI;
+
+/**
+ * set the width of the line
+ * @param {number} w
+ */
+const width = w => {
+   w = +w;
+   _turtle.width = w;
+   _imageCtx.lineWidth = w;
+};
+
+/**
+ * write some text at the turtle position, with custom or default font.
+ *
+ * ideally we'd like this to rotate the text based on
+ * the turtle orientation, but this will require some clever
+ * canvas transformations which aren't implemented yet.
+ * @param {string} msg
+ * @param {string} font
+ */
+const write = (msg, font) => {
+   const {x, y} = _turtle.pos;
+
+   _imageCtx.save();
+   _centerCoords(_imageCtx);
+
+   //_imageCtx.rotate(turtle.angle);
+   _imageCtx.translate(x, y);
+   _imageCtx.transform(1, 0, 0, -1, 0, 0);
+   _imageCtx.translate(-x, -y);
+   _imageCtx.font = font;
+   _imageCtx.fillText(msg, x, y);
+   _imageCtx.restore();
+
+   drawIf();
+};
+
+/**
+ * set the turtle draw shape
+ *
+ * currently supports triangle (default), circle, square, and turtle
+ * @param {string} s
+ */
+const shape = s => {
+   _turtle.shape = s;
+   draw();
+};
+
+/**
+ * set background color using RGB values in the range 0 - 255.
+ *
+ * currently, this has the unintended effect of wiping the entire canvas.
+ * (to-do: fix later)
+ * @param {number} r
+ * @param {number} g
+ * @param {number} b
+ * @param {number} a alpha
+ */
+const bgColour = (r, g, b, a) => {
+   const {width: w, height: h} = _imageCanvas;
+   _imageCtx.fillStyle = `rgba(${r}, ${g}, ${b}, ${a})`;
+   _imageCtx.fillRect(0, 0, w, h);
+};
+const bgColor = bgColour;
+
+/**
+ * set line color using RGB values in the range 0 - 255.
+ * @param {number} r
+ * @param {number} g
+ * @param {number} b
+ * @param {number} a alpha
+ */
+const colour = (r, g, b, a) => {
+   r = +r;
+   g = +g;
+   b = +b;
+   a = +a;
+
+   _imageCtx.strokeStyle = `rgba(${r}, ${g}, ${b}, ${a})`;
+   const c = _turtle.colour;
+   c.r = r;
+   c.g = g;
+   c.b = b;
+   c.a = a;
+};
+const color = colour;
+
+/**
+ * Returns a pseudo-random integer in the range `min` <= n <= `max` (inclusive)
+ * @param {number} min
+ * @param {number} max
+ */
+const random = (min, max) => {
+   min = +min;
+   max = +max;
+
+   return Math.floor(Math.random() * (max - min + 1) + min);
+};
+
+/**
+ * repeatedly run an "action" callback `n` times
+ * @param {number} n (integer)
+ * @param {Function} action (callback)
+ */
+const repeat = (n, action) => { while (n-- > 0) action(); };
+
+/**
+ * `setInterval` alias, but 2-adic (no rest args)
+ * @param {TimerHandler} f
+ * @param {number | undefined} ms
+ */
+const animate = (f, ms) => setInterval(f, ms);
+
+/**
+ * main program/script, mostly UI code.
+ *
+ * this fn is used to encapsulate private stuff that the user shouldn't access
+ */
+const _main = () => {
+   const doc = document;
+
+   /**
+    * check if value is of type `number`, and an unsigned 32bit integer
+    * @param {*} x value to check
+   */
+   const is_u32 = x => typeof x == 'number' && x == (x >>> 0);
```

Comment:
B/c comparison operator `==` has lower priority than unsigned right-shift operator `>>>`, parens `()` are unneeded:
`const is_u32 = x => typeof x == 'number' && x == x >>> 0;`

Reply:
True. And I think I should rename it to match the typedef

Reply:
Holup I discovered a "bug"
```js
is_u32(-0) //true
// lmao
```

Only floats have negative zero, this is wrong